File: Backend/pdks-backend/collector.js
const mqtt = require('mqtt');
const pool = require('./db');

const mqttHost = process.env.MQTT_HOST || '127.0.0.1';
const client = mqtt.connect(`mqtt://${mqttHost}:1883`);

// --- TRANSLATION MAPS ---
// Converts ESP32 string payloads back into SMALLINT for PostgreSQL
const mapResult = { 'granted': 0, 'unknown': 1, 'expired': 2, 'schedule': 3, 'manual': 4 };
const mapDir = { 'in': 0, 'out': 1 };
const mapMode = { 'online': 0, 'offline': 1 };
const mapTsrc = { 'ntp': 0, 'rtc': 1, 'invalid': 2 };

client.on('connect', () => {
    console.log('Collector connected to Mosquitto MQTT Broker');
    client.subscribe('pdks/merkez/dev/+/event', { qos: 1 });
    client.subscribe('pdks/merkez/dev/+/status', { qos: 1 });
    client.subscribe('pdks/merkez/dev/+/hb', { qos: 0 }); 
    client.subscribe('pdks/merkez/dev/+/cmd/res', { qos: 1 });
});

client.on('error', (err) => {
    console.error('MQTT connection error:', err.message);
});

client.on('message', async (topic, message) => {
    await pool.ready; // Wait for db.js to build tables

    const payload = message.toString();
    const topicParts = topic.split('/');
    const deviceId = topicParts[3]; 

    // --- Process Door Scans (Events) ---
    if (topic.endsWith('/event')) {
        let data;
        try {
            data = JSON.parse(payload);
        } catch (e) {
            console.error("Invalid JSON payload");
            return;
        }

        const now = Math.floor(Date.now() / 1000);

        // Opportunistic presence/firmware tracking - "fw" only ever appears on
        // event payloads, never on heartbeats, so this is the only place it can
        // be captured. Runs regardless of whether the insert below turns out to
        // be a duplicate, since it's just presence + version, not the event itself.
        try {
            await pool.query(
                `INSERT INTO devices (id, durum, son_gorulme, fw)
                 VALUES ($1, 'online', $2, $3)
                 ON CONFLICT(id) DO UPDATE SET durum = 'online', son_gorulme = EXCLUDED.son_gorulme, fw = EXCLUDED.fw`,
                [deviceId, now, data.fw || null]
            );
        } catch (err) {
            console.error('Device fw/presence update error:', err);
        }
        
        // Translate strings to integers (defaulting to safe values if undefined)
        const resInt = mapResult[data.res] ?? 1; // Default: 1 (unknown)
        const dirInt = mapDir[data.dir] ?? 0;    // Default: 0 (in)
        const modeInt = mapMode[data.mode] ?? 0; // Default: 0 (online)
        const tsrcInt = mapTsrc[data.tsrc] ?? 2; // Default: 2 (invalid)

        // Find the employee_id associated with this UID
        let employeeId = null;
        try {
            const cardLookup = await pool.query('SELECT employee_id FROM cards WHERE uid = $1', [data.uid]);
            if (cardLookup.rows.length > 0) {
                employeeId = cardLookup.rows[0].employee_id;
            }
        } catch (err) {
            console.error('Error looking up employee ID:', err);
        }
        
        // Insert all columns including dir, mode, ts_source, and employee_id
        const query = `
            INSERT INTO access_events 
            (device_id, seq, uid, employee_id, ts_utc, ts_source, dir, result, mode, alindi_at) 
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
        `;
        const values = [deviceId, data.seq, data.uid, employeeId, data.ts, tsrcInt, dirInt, resInt, modeInt, now];

        try {
            await pool.query(query, values);
            console.log(`Record saved. Sending ACK for Seq: ${data.seq}`);
            sendAck(deviceId, data.seq);
        } catch (err) {
            if (err.code === '23505') {
                console.log(`Duplicate record ignored (Device: ${deviceId}, Seq: ${data.seq})`);
                sendAck(deviceId, data.seq);
            } else {
                console.error('Database insertion error:', err);
            }
        }
    } 
    
    // --- Process Online/Offline Status (LWT) ---
    else if (topic.endsWith('/status')) {
        const now = Math.floor(Date.now() / 1000);
        const query = `
            INSERT INTO devices (id, durum, son_gorulme) 
            VALUES ($1, $2, $3)
            ON CONFLICT(id) DO UPDATE SET durum = EXCLUDED.durum, son_gorulme = EXCLUDED.son_gorulme
        `;
        
        try {
            await pool.query(query, [deviceId, payload, now]);
            console.log(`Device ${deviceId} status: ${payload}`);
        } catch (err) {
            console.error('Device status update error:', err);
        }
    }

    // --- Process Heartbeats (hb) ---
    else if (topic.endsWith('/hb')) {
        const now = Math.floor(Date.now() / 1000);
        let hb = {};
        try {
            hb = JSON.parse(payload);
        } catch (e) {
            console.error('Invalid heartbeat JSON, storing presence only.');
        }

        // Upsert (not a blind UPDATE) so a heartbeat that happens to arrive
        // before this device's first /status message doesn't just silently
        // touch zero rows and lose the data.
        const query = `
            INSERT INTO devices (id, durum, son_gorulme, queue_depth, heap_free, queue_overflow, uptime_s)
            VALUES ($1, 'online', $2, $3, $4, $5, $6)
            ON CONFLICT(id) DO UPDATE SET
                durum = 'online',
                son_gorulme = EXCLUDED.son_gorulme,
                queue_depth = EXCLUDED.queue_depth,
                heap_free = EXCLUDED.heap_free,
                queue_overflow = EXCLUDED.queue_overflow,
                uptime_s = EXCLUDED.uptime_s
        `;
        try {
            await pool.query(query, [
                deviceId, now,
                hb.queue ?? null,
                hb.heap ?? null,
                hb.qOverflow ?? null,
                hb.uptime ?? null,
            ]);
        } catch (err) {
            console.error('Heartbeat update error:', err);
        }
    }

    // --- Process Command Responses ---
    else if (topic.endsWith('/cmd/res')) {
        console.log(`[CMD RESULT] Device ${deviceId} responded: ${payload}`);
    }
});

function sendAck(deviceId, seqNumber) {
    const ackTopic = `pdks/merkez/dev/${deviceId}/event/ack`;
    const ackPayload = JSON.stringify({ ack_seq: seqNumber });
    
    client.publish(ackTopic, ackPayload, { qos: 1 }, (err) => {
        if (err) console.error('Failed to send ACK:', err);
    });
}

File: backend-django/collector/collector.py
"""
collector.js'in paho-mqtt + psycopg2 kullanan birebir portu. Aynı topic'lere
subscribe olur, aynı string->SMALLINT çeviri tablolarını ve zaman damgası
mantık kontrolünü uygular, aynı tablolara yazar.
"""
import json
import os
import time

import paho.mqtt.client as mqtt
import psycopg2

import db

MQTT_HOST = os.environ.get("MQTT_HOST", "127.0.0.1")
MQTT_PORT = int(os.environ.get("MQTT_PORT", "1883"))

# --- ÇEVİRİ TABLOLARI ---
# ESP32'nin string payload'larını PostgreSQL için SMALLINT'e çevirir
MAP_RESULT = {"granted": 0, "unknown": 1, "expired": 2, "schedule": 3, "manual": 4}
MAP_DIR = {"in": 0, "out": 1}
MAP_MODE = {"online": 0, "offline": 1}
MAP_TSRC = {"ntp": 0, "rtc": 1, "invalid": 2}

conn = db.connect()  # returns an autocommit connection - see db.py
db.wait_for_schema(conn)


def now_s():
    return int(time.time())


def send_ack(client, device_id, seq_number):
    ack_topic = f"pdks/merkez/dev/{device_id}/event/ack"
    ack_payload = json.dumps({"ack_seq": seq_number})
    try:
        client.publish(ack_topic, ack_payload, qos=1)
    except Exception as err:
        print(f"Failed to send ACK: {err}", flush=True)


def handle_event(client, device_id, payload):
    try:
        data = json.loads(payload)
    except json.JSONDecodeError:
        print("Invalid JSON payload", flush=True)
        return

    now = now_s()

    # "fw" sadece event payload'larında görünür (heartbeat'te değil), burası tek yakalama noktası.
    try:
        with conn.cursor() as cur:
            cur.execute(
                """
                INSERT INTO devices (id, status, last_seen_at, fw, created_at, updated_at)
                VALUES (%s, 'online', %s, %s, NOW(), NOW())
                ON CONFLICT(id) DO UPDATE SET status = 'online', last_seen_at = EXCLUDED.last_seen_at, fw = EXCLUDED.fw, updated_at = NOW()
                """,
                (device_id, now, data.get("fw") or None),
            )
    except Exception as err:
        print(f"Device fw/presence update error: {err}", flush=True)

    # String'leri integer'a çevir (tanımsızsa güvenli değerlere düş)
    res_int = MAP_RESULT.get(data.get("res"), 1)   # Varsayılan: 1 (unknown)
    dir_int = MAP_DIR.get(data.get("dir"), 0)      # Varsayılan: 0 (in)
    mode_int = MAP_MODE.get(data.get("mode"), 0)   # Varsayılan: 0 (online)
    tsrc_int = MAP_TSRC.get(data.get("tsrc"), 2)   # Varsayılan: 2 (invalid)

    ts = data.get("ts")
    # Mantık kontrolü: zaman damgası gelecekte ise (>10 dk) ya da 2025'ten önceyse geçersiz say
    if ts is None or ts > (now + 600) or ts < 1735689600:
        print(f"[TIMESTAMP ANOMALY] Device {device_id} sent suspect ts: {ts}. Overriding tsrc to invalid.", flush=True)
        tsrc_int = 2  # TSRC_INVALID

    # Bu UID'ye bağlı employee_id'yi bul
    employee_id = None
    try:
        with conn.cursor() as cur:
            cur.execute("SELECT employee_id FROM cards WHERE uid = %s", (data.get("uid"),))
            row = cur.fetchone()
            if row:
                employee_id = row[0]
    except Exception as err:
        print(f"Error looking up employee ID: {err}", flush=True)

    query = """
        INSERT INTO access_events
        (device_id, seq, uid, employee_id, ts_utc, ts_source, dir, result, mode, ingested_at, raw_payload)
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s::jsonb)
    """
    values = (  # raw_payload replay/backfill için saklanır, eşleme hatası veri kaybettirmesin diye
        device_id, data.get("seq"), data.get("uid"), employee_id, ts, tsrc_int, dir_int, res_int, mode_int, now,
        json.dumps(data),
    )

    try:
        with conn.cursor() as cur:
            cur.execute(query, values)
        print(f"Record saved. Sending ACK for Seq: {data.get('seq')}", flush=True)
        send_ack(client, device_id, data.get("seq"))
    except psycopg2.errors.UniqueViolation:
        conn.rollback()
        print(f"Duplicate record ignored (Device: {device_id}, Seq: {data.get('seq')})", flush=True)
        send_ack(client, device_id, data.get("seq"))
    except Exception as err:
        conn.rollback()
        print(f"Database insertion error: {err}", flush=True)


def handle_status(device_id, payload):
    now = now_s()
    query = """
        INSERT INTO devices (id, status, last_seen_at, created_at, updated_at)
        VALUES (%s, %s, %s, NOW(), NOW())
        ON CONFLICT(id) DO UPDATE SET status = EXCLUDED.status, last_seen_at = EXCLUDED.last_seen_at, updated_at = NOW()
    """
    try:
        with conn.cursor() as cur:
            cur.execute(query, (device_id, payload, now))
        print(f"Device {device_id} status: {payload}", flush=True)
    except Exception as err:
        conn.rollback()
        print(f"Device status update error: {err}", flush=True)


def handle_heartbeat(device_id, payload):
    now = now_s()
    try:
        hb = json.loads(payload)
    except json.JSONDecodeError:
        hb = {}
        print("Invalid heartbeat JSON, storing presence only.", flush=True)

    # upsert, kör UPDATE değil - ilk /status'tan önce gelen heartbeat veri kaybettirmesin
    query = """
        INSERT INTO devices (id, status, last_seen_at, queue_depth, heap_free, queue_overflow, uptime_s, created_at, updated_at)
        VALUES (%s, 'online', %s, %s, %s, %s, %s, NOW(), NOW())
        ON CONFLICT(id) DO UPDATE SET
            status = 'online',
            last_seen_at = EXCLUDED.last_seen_at,
            queue_depth = EXCLUDED.queue_depth,
            heap_free = EXCLUDED.heap_free,
            queue_overflow = EXCLUDED.queue_overflow,
            uptime_s = EXCLUDED.uptime_s,
            updated_at = NOW()
    """
    try:
        with conn.cursor() as cur:
            cur.execute(
                query,
                (
                    device_id, now,
                    hb.get("queue"), hb.get("heap"), hb.get("qOverflow"), hb.get("uptime"),
                ),
            )
    except Exception as err:
        conn.rollback()
        print(f"Heartbeat update error: {err}", flush=True)


def handle_cmd_res(device_id, payload):
    print(f"[CMD RESULT] Device {device_id} responded: {payload}", flush=True)

    # Sadece OTA yanıtları kalıcı saklanır (panel ilerleme göstersin diye), diğerleri geçici ack.
    if payload.startswith("ota_"):
        now = now_s()
        try:
            with conn.cursor() as cur:
                cur.execute(
                    """
                    INSERT INTO devices (id, status, last_seen_at, ota_status, ota_updated_at, created_at, updated_at)
                    VALUES (%s, 'online', %s, %s, %s, NOW(), NOW())
                    ON CONFLICT(id) DO UPDATE SET
                        status = 'online',
                        last_seen_at = EXCLUDED.last_seen_at,
                        ota_status = EXCLUDED.ota_status,
                        ota_updated_at = EXCLUDED.ota_updated_at,
                        updated_at = NOW()
                    """,
                    (device_id, now, payload, now),
                )
        except Exception as err:
            conn.rollback()
            print(f"OTA status update error: {err}", flush=True)


def on_connect(client, userdata, flags, rc):
    print("Collector connected to Mosquitto MQTT Broker", flush=True)
    client.subscribe("pdks/merkez/dev/+/event", qos=1)
    client.subscribe("pdks/merkez/dev/+/status", qos=1)
    client.subscribe("pdks/merkez/dev/+/hb", qos=0)
    client.subscribe("pdks/merkez/dev/+/cmd/res", qos=1)


def on_disconnect(client, userdata, rc):
    print(f"MQTT connection error/disconnected: rc={rc}", flush=True)


def on_message(client, userdata, msg):
    topic = msg.topic
    payload = msg.payload.decode("utf-8", errors="replace")
    topic_parts = topic.split("/")
    device_id = topic_parts[3] if len(topic_parts) > 3 else None

    if topic.endswith("/event"):
        handle_event(client, device_id, payload)
    elif topic.endswith("/status"):
        handle_status(device_id, payload)
    elif topic.endswith("/hb"):
        handle_heartbeat(device_id, payload)
    elif topic.endswith("/cmd/res"):
        handle_cmd_res(device_id, payload)


def main():
    client = mqtt.Client()
    client.on_connect = on_connect
    client.on_disconnect = on_disconnect
    client.on_message = on_message

    while True:
        try:
            client.connect(MQTT_HOST, MQTT_PORT, keepalive=60)
            client.loop_forever()
        except Exception as err:
            print(f"MQTT connection error: {err}", flush=True)
            time.sleep(3)


if __name__ == "__main__":
    main()
